design/metadata_kv/index.md

```diff
@@ -0,0 +1,187 @@
+        groupIds = append(groupId, membership.GroupID)
+    }
+    ...
+    return groupIds
+}
+
+```
+
+It is possible to create a 2-way index for many-to-many relationships, but this is not generally required - it would be simpler to simply do a scan when needed and filter only the relevant values.
+This is because the IAM keyspace is relatively very small - it would be surprising if the biggest lakeFS installation to ever exist would contain more than, say, 50k users.
+
+Some care does need to be applied when managing these secondary indices - for example, when deleting an entity, secondary indices need to be pruned first, to avoid inconsistencies. 
+
+### Graveler Metadata - Commits, Tags, Repositories
+
+These are simpler entities - commits and tags are immutable and could potentially be stored on the object store and not in the key/value store (they are cheap either way, so it doesn't make much of a difference).
+Repositories and tags, are also returned in lexicographical order, which map well to the suggested abstraction. Commits are usually returned using parent traversal, so no scanning takes place anyway. 
+
+There are no special concurrency requirements for these entities, apart for last-write-wins which is already the case for all modern stores.
+
+### Graveler Metadata - Branches and Staged Writes
+
+This is where concurrency control gets interesting, and where lakeFS is expected to provide a **correct** system whose semantics are well understood (lakeFS currently [falls short](https://github.com/treeverse/lakeFS/issues/2405) in that regard).
+
+Concurrency is more of an issue here because of how a commit works: when a commit starts, it scans the currently staged changes, applies them to the current commit pointed to by the branch, updating the branch reference and removing the staged changes it applied.
+
+Getting this right means we have to take care of the following:
+
+1. Ensure all staged changes that finished successfully before the commit started are applied as part of the commit (causality)
+1. Ensure acknowledged writes end up either in a resulting commit, or staged to be committed (no lost writes)
+
+To do this, we will employ 2 mechanisms:
+
+1. [Optimistic Concurrency Control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) on the branch pointer using `SetIf()`
+1. Reliance on write [idempotency](https://en.wikipedia.org/wiki/Idempotence) provided by Graveler (i.e., Writing the same exact entry, with the same identity - will not appear as a change)
+
+This is what the proposed implementation will look like:
+
+#### Committer flow
+
+We add an additional field to each `Branch` object: In addition to the existing `staging_token`, we add an array of strings named `sealed_tokens`.
+
+1. get branch, find current `staging_token`
+1. use `SetIf()` to update the branch (if not modified by another process): push existing `staging_token` into `sealed_tokens`, set new uuid as `staging_token`.
```

Comment:
IIUC this is `SetIf` the (existing, pushed onto `staging_token`) `staging_token` is unchanged?

Reply:
Made it clear that `SetIf()` is actually a CAS operation on the entire branch (`staging_token`, `sealed_tokens` and `commit_id`).